src/bootloader.rs

```diff
@@ -0,0 +1,500 @@
+        //   3. no options line (append the entire thing)
+        for line in &mut self.lines {
+            if let Some(cmdline) = strip_ble_key(line, "options") {
+                let segment = split_cmdline(cmdline).find(|s| s.starts_with(key));
+
+                if let Some(old) = segment {
+                    // 1. Replace existing key
+                    let range = substr_range(line, old).unwrap();
+                    line.replace_range(range, arg);
+                } else {
+                    // 2. Append new argument
+                    line.push(' ');
+                    line.push_str(arg);
+                }
+
+                return;
+            }
+        }
+
+        // 3. Append new "options" line with our argument
+        self.lines.push(format!("options {arg}"));
+    }
+
+    /// Adjusts the kernel command-line arguments by adding a composefs= parameter (if appropriate)
+    /// and adding additional arguments, as requested.
+    pub fn adjust_cmdline(&mut self, composefs: Option<&str>, extra: &[&str]) {
+        if let Some(id) = composefs {
+            self.add_cmdline(&format!("composefs={id}"));
+        }
+
+        for item in extra {
+            self.add_cmdline(item);
+        }
+    }
+}
+
+// TODO: put this in a better spot
+// Reminder: it only makes sense to do boot stuff on a legit repository because we need to read the
+// boot loader entry content.  But: wouldn't it be nice to use the trait convenience wrappers
+// anyway?
+fn read_file<ObjectID: FsVerityHashValue>(
+    file: &RegularFile<ObjectID>,
+    repo: &Repository<ObjectID>,
+) -> Result<Box<[u8]>> {
+    match file {
+        RegularFile::Inline(data) => Ok(data.clone()),
+        RegularFile::External(id, size) => {
+            // TODO: mmap niceness?
+            // TODO: verify size: it could be faked
+            let mut data = Vec::with_capacity(*size as usize);
+            // TODO: read loop...
+            read(repo.open_object(id)?, spare_capacity(&mut data)).unwrap();
+            Ok(Box::from(data))
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Type1Entry<ObjectID: FsVerityHashValue> {
+    /// This is the basename of the bootloader entry .conf file
+    pub filename: Box<OsStr>,
+    pub entry: BootLoaderEntryFile,
+    pub files: HashMap<Box<str>, RegularFile<ObjectID>>,
+}
+
+impl<ObjectID: FsVerityHashValue> Type1Entry<ObjectID> {
+    // Relocates boot resources.
+    //
+    // This is a bit of a strange operation: for each file mentioned in the bootloader entry, move
+    // the file into the given 'entry_id' pathname and rename the entry file itself to
+    // "{entry_id}.conf".
+    pub fn relocate(&mut self, entry_id: &str) {
+        self.filename = Box::from(format!("{entry_id}.conf").as_ref());
+        for line in &mut self.entry.lines {
+            for key in ["linux", "initrd", "efi"] {
+                if let Some(value) = strip_ble_key(line, key) {
```

Comment:
The nesting is pretty deep, around this point I tend to do e.g.

`let Some(value) = strip_ble_key(line, key) else { continue };`